src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingProcessor.cs

```diff
@@ -0,0 +1,418 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Core;
+using Azure.Identity;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+
+// description
+// - Get the number of total processors and total partitions
+// - Calculate how many partitions should be assigned to each processor
+// - Lock the partition(s)
+// - Pass the partition ids into the StaticPartitionProcessor and run until the number of total processors changes
+
+// usage
+//  var cts = new CancellationTokenSource();
+//  var processor = new PartitionLockingProcessor();
+//  await processor.RunProcessor(cts);
+
+// todos:
+// - detect if a partition has no ownership updates for some time
+// - handle cases where the ratio of processors to partitions is not divided evenly
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingProcessor
+    {
+        private string _processorId;
+
+        private List<int> _ownedPartitions;
+
+        private BlobContainerClient _containerClient;
+
+        private int _totalProcessorsRunning = 0;
+
+        // consider injecting these or querying event hub for max partitions
+        private int _maximumProcessors = 8;
+
+        private int _maximumPartitions = 8;
+
+        // settings that should be passed in as environment variables
+        private Uri _paritionManagerUri = new Uri("<blob uri goes here>");
```

Comment:
Are you looking for sign off on the overall concept and then planning on doing a pass to wire these up before committing?

Reply:
Yeah mainly wanted to stop here and validate the concept before proceeding with additional items like dependency injection, logging, production readiness etc so that if we prefer to go another direction then it wouldn't be throw away work.